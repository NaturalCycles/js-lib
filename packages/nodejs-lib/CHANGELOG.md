See [Github Releases](https://github.com/NaturalCycles/nodejs-lib/releases)
