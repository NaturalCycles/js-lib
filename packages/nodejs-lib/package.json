{
  "name": "@naturalcycles/nodejs-lib",
  "type": "module",
  "version": "14.3.0",
  "scripts": {
    "build": "dev-lib build",
    "test": "dev-lib test",
    "lint": "dev-lib lint",
    "bt": "dev-lib bt",
    "lbt": "dev-lib lbt",
    "slack-this-debug": "tsx ./src/bin/slack-this.ts --channel test --msg 'Hello slack!'",
    "secrets-gen-key-debug": "tsx ./src/bin/secrets-gen-key.ts",
    "secrets-encrypt-debug": "tsx ./src/bin/secrets-encrypt.ts",
    "secrets-decrypt-debug": "tsx ./src/bin/secrets-decrypt.ts",
    "kpy-debug": "tsx ./src/bin/kpy.ts --verbose scripts tmp/scripts",
    "kpy-debug2": "tsx ./src/bin/kpy.ts --verbose scripts bench non-ex non-ex/** colors* tmp/scripts",
    "kpy-debug3": "tsx ./src/bin/kpy.ts --verbose src colors csv stream non-ex non-ex/** tmp/src",
    "json2env-debug": "tsx ./src/bin/json2env.ts ./src/test/someFile.json"
  },
  "dependencies": {
    "@naturalcycles/js-lib": "workspace:^15",
    "@types/js-yaml": "^4",
    "@types/jsonwebtoken": "^9",
    "@types/yargs": "^16",
    "ajv": "^8",
    "ajv-formats": "^3",
    "ajv-keywords": "^5",
    "chalk": "^5",
    "dotenv": "^16",
    "joi": "^17",
    "js-yaml": "^4",
    "jsonwebtoken": "^9",
    "lru-cache": "^11",
    "through2-concurrent": "^2",
    "tinyglobby": "^0.2",
    "yargs": "^17"
  },
  "devDependencies": {
    "@naturalcycles/dev-lib": "^18",
    "@types/through2-concurrent": "^2"
  },
  "bin": {
    "kpy": "dist/bin/kpy.js",
    "json2env": "dist/bin/json2env.js",
    "generate-build-info": "dist/bin/generate-build-info.js",
    "slack-this": "dist/bin/slack-this.js",
    "secrets-gen-key": "dist/bin/secrets-gen-key.js",
    "secrets-encrypt": "dist/bin/secrets-encrypt.js",
    "secrets-decrypt": "dist/bin/secrets-decrypt.js"
  },
  "files": [
    "dist",
    "src",
    "!src/test",
    "!src/**/*.test.ts",
    "!src/**/__snapshots__",
    "!src/**/__exclude"
  ],
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git@github.com:NaturalCycles/js-libs.git",
    "directory": "packages/nodejs-lib"
  },
  "engines": {
    "node": ">=22.12.0"
  },
  "description": "Standard library for Node.js",
  "author": "Natural Cycles Team",
  "license": "MIT",
  "envByBranch": {
    "master": "master",
    "*": "branch"
  }
}
