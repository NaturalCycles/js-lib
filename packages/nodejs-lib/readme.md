## @naturalcycles/nodejs-lib

> Standard library for Node.js

[![npm](https://img.shields.io/npm/v/@naturalcycles/nodejs-lib/latest.svg)](https://www.npmjs.com/package/@naturalcycles/nodejs-lib)
[![publish size](https://badgen.net/packagephobia/publish/@naturalcycles/nodejs-lib)](https://packagephobia.now.sh/result?p=@naturalcycles/nodejs-lib)
[![install size](https://badgen.net/packagephobia/install/@naturalcycles/nodejs-lib)](https://packagephobia.now.sh/result?p=@naturalcycles/nodejs-lib)
[![Maintainability](https://api.codeclimate.com/v1/badges/119a3b4735c4ed81cf84/maintainability)](https://codeclimate.com/github/NaturalCycles/nodejs-lib/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/119a3b4735c4ed81cf84/test_coverage)](https://codeclimate.com/github/NaturalCycles/nodejs-lib/test_coverage)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![Actions](https://github.com/NaturalCycles/nodejs-lib/workflows/ci/badge.svg)](https://github.com/NaturalCycles/nodejs-lib/actions)

# [Documentation](https://naturalcycles.github.io/nodejs-lib/)

# Packaging

- `engines.node`: Latest Node.js LTS
- `main: dist/index.js`: commonjs, es2020
- `types: dist/index.d.ts`: typescript types
- `/src` folder with source `*.ts` files included
