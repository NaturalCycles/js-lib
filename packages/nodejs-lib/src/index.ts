import Ajv from 'ajv'
export { glob, globSync } from 'tinyglobby'
import type {
  AlternativesSchema,
  AnySchema,
  ArraySchema,
  BinarySchema,
  BooleanSchema,
  DateSchema,
  FunctionSchema,
  ObjectSchema,
  ValidationErrorItem,
} from 'joi'
export * from './buffer/buffer.util.js'
export * from './colors/colors.js'
export * from './csv/csvReader.js'
export * from './csv/csvWriter.js'
export * from './csv/transformToCSV.js'
export * from './diff/tableDiff.js'
export * from './fs/fs2.js'
export * from './fs/json2env.js'
export * from './fs/kpy.js'
export * from './infra/process.util.js'
export * from './jwt/jwt.service.js'
export * from './log/log.util.js'
export * from './script/runScript.js'
export * from './security/crypto.util.js'
export * from './security/hash.util.js'
export * from './security/id.util.js'
export * from './security/nanoid.js'
export * from './security/secret.util.js'
export * from './slack/slack.service.js'
export * from './slack/slack.service.model.js'
export * from './stream/ndjson/ndjson.model.js'
export * from './stream/ndjson/ndjsonMap.js'
export * from './stream/ndjson/ndjsonStreamForEach.js'
export * from './stream/ndjson/transformJsonParse.js'
export * from './stream/ndjson/transformToNDJson.js'
export * from './stream/pipeline/pipeline.js'
export * from './stream/progressLogger.js'
export * from './stream/readable/readableCreate.js'
export * from './stream/readable/readableForEach.js'
export * from './stream/readable/readableFromArray.js'
export * from './stream/readable/readableToArray.js'
export * from './stream/stream.model.js'
export * from './stream/transform/transformChunk.js'
export * from './stream/transform/transformFilter.js'
export * from './stream/transform/transformLimit.js'
export * from './stream/transform/transformLogProgress.js'
export * from './stream/transform/transformMap.js'
export * from './stream/transform/transformMapSimple.js'
export * from './stream/transform/transformMapSync.js'
export * from './stream/transform/transformNoOp.js'
export * from './stream/transform/transformOffset.js'
export * from './stream/transform/transformSplit.js'
export * from './stream/transform/transformTap.js'
export * from './stream/transform/transformTee.js'
export * from './stream/transform/transformThrottle.js'
export * from './stream/transform/transformToArray.js'
export * from './stream/transform/worker/baseWorkerClass.js'
export * from './stream/transform/worker/transformMultiThreaded.js'
export * from './stream/transform/worker/transformMultiThreaded.model.js'
export * from './stream/writable/writableForEach.js'
export * from './stream/writable/writableFork.js'
export * from './stream/writable/writablePushToArray.js'
export * from './stream/writable/writableVoid.js'
export * from './string/inspect.js'
export * from './util/buildInfo.util.js'
export * from './util/env.util.js'
export * from './util/exec2.js'
export * from './util/git2.js'
export * from './util/lruMemoCache.js'
export * from './util/zip.util.js'
export * from './validation/ajv/ajv.util.js'
export * from './validation/ajv/ajvSchema.js'
export * from './validation/ajv/ajvValidationError.js'
export * from './validation/ajv/getAjv.js'
export * from './validation/joi/joi.extensions.js'
export * from './validation/joi/joi.model.js'
export * from './validation/joi/joi.shared.schemas.js'
export * from './validation/joi/joi.validation.error.js'
export * from './validation/joi/joi.validation.util.js'
export type { StringSchema } from './validation/joi/string.extensions.js'
export * from './yargs.util.js'

export type {
  AlternativesSchema,
  AnySchema,
  ArraySchema,
  BinarySchema,
  BooleanSchema,
  DateSchema,
  FunctionSchema,
  ObjectSchema,
  ValidationErrorItem,
  // extended
  // NumberSchema,
  // StringSchema,
}

export { Ajv }
