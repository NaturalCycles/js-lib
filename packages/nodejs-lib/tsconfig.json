{
  "extends": "@naturalcycles/dev-lib/cfg/tsconfig.json",
  "compilerOptions": {
    // "erasableSyntaxOnly": true,
  },
  "include": ["src"],
  "exclude": ["**/__exclude"],
}
