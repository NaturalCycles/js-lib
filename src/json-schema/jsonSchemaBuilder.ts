import { _uniq } from '../array/array.util'
import type {
  AnyObject,
  BaseDBEntity,
  JsonSchemaAllOf,
  JsonSchemaArray,
  JsonSchemaOneOf,
  JsonSchemaTuple,
} from '../index'
import { _deepCopy, _sortObject, mergeJsonSchemaObjects } from '../index'
import { JSON_SCHEMA_ORDER } from './jsonSchema.cnst'
import type {
  JsonSchema,
  JsonSchemaAny,
  JsonSchemaBoolean,
  JsonSchemaConst,
  JsonSchemaEnum,
  JsonSchemaNull,
  JsonSchemaNumber,
  JsonSchemaObject,
  JsonSchemaRef,
  JsonSchemaString,
} from './jsonSchema.model'

/* eslint-disable id-blacklist, @typescript-eslint/explicit-module-boundary-types */

export interface JsonSchemaBuilder<T = unknown> {
  build: () => JsonSchema<T>
}

/**
 * Fluent (chainable) API to manually create Json Schemas.
 * Inspired by Joi
 */
export const jsonSchema = {
  any<T = unknown>() {
    return new JsonSchemaAnyBuilder<T, JsonSchemaAny<T>>({})
  },
  const<T = unknown>(value: T) {
    return new JsonSchemaAnyBuilder<T, JsonSchemaConst<T>>({
      const: value,
    })
  },
  null() {
    return new JsonSchemaAnyBuilder<null, JsonSchemaNull>({
      type: 'null',
    })
  },
  ref<T = unknown>($ref: string) {
    return new JsonSchemaAnyBuilder<T, JsonSchemaRef<T>>({
      $ref,
    })
  },
  enum<T = unknown>(enumValues: T[]) {
    return new JsonSchemaAnyBuilder<T, JsonSchemaEnum<T>>({ enum: enumValues })
  },
  boolean() {
    return new JsonSchemaAnyBuilder<boolean, JsonSchemaBoolean>({
      type: 'boolean',
    })
  },
  buffer() {
    return new JsonSchemaAnyBuilder<Buffer, JsonSchemaAny<Buffer>>({
      instanceof: 'Buffer',
    })
  },

  // number types
  number() {
    return new JsonSchemaNumberBuilder()
  },
  integer() {
    return new JsonSchemaNumberBuilder().integer()
  },
  unixTimestamp() {
    return new JsonSchemaNumberBuilder().unixTimestamp()
  },
  unixTimestamp2000() {
    return new JsonSchemaNumberBuilder().unixTimestamp2000()
  },
  // string types
  string() {
    return new JsonSchemaStringBuilder()
  },
  dateString() {
    return new JsonSchemaStringBuilder().date()
  },
  // email: () => new JsonSchemaStringBuilder().email(),
  // complex types
  object<T extends AnyObject>(props: {
    [k in keyof T]: JsonSchemaAnyBuilder<T[k]>
  }) {
    return new JsonSchemaObjectBuilder<T>().addProperties(props)
  },
  rootObject<T extends AnyObject>(props: {
    [k in keyof T]: JsonSchemaAnyBuilder<T[k]>
  }) {
    return new JsonSchemaObjectBuilder<T>().addProperties(props).$schemaDraft7()
  },
  array<ITEM = unknown>(itemSchema: JsonSchemaAnyBuilder<ITEM>) {
    return new JsonSchemaArrayBuilder<ITEM>(itemSchema)
  },
  tuple<T extends any[] = unknown[]>(items: JsonSchemaAnyBuilder[]) {
    return new JsonSchemaTupleBuilder<T>(items)
  },
  oneOf<T = unknown>(items: JsonSchemaAnyBuilder[]) {
    return new JsonSchemaAnyBuilder<T, JsonSchemaOneOf<T>>({
      oneOf: items.map(b => b.build()),
    })
  },
  allOf<T = unknown>(items: JsonSchemaAnyBuilder[]) {
    return new JsonSchemaAnyBuilder<T, JsonSchemaAllOf<T>>({
      allOf: items.map(b => b.build()),
    })
  },
}

export class JsonSchemaAnyBuilder<T = unknown, SCHEMA_TYPE extends JsonSchema<T> = JsonSchema<T>>
  implements JsonSchemaBuilder<T>
{
  constructor(protected schema: SCHEMA_TYPE) {}

  /**
   * Used in ObjectBuilder to access schema.optionalProperty
   */
  getSchema(): SCHEMA_TYPE {
    return this.schema
  }

  $schema($schema: string): this {
    Object.assign(this.schema, { $schema })
    return this
  }

  $schemaDraft7(): this {
    this.$schema('http://json-schema.org/draft-07/schema#')
    return this
  }

  $id($id: string): this {
    Object.assign(this.schema, { $id })
    return this
  }

  title(title: string): this {
    Object.assign(this.schema, { title })
    return this
  }

  description(description: string): this {
    Object.assign(this.schema, { description })
    return this
  }

  deprecated(deprecated = true): this {
    Object.assign(this.schema, { deprecated })
    return this
  }

  type(type: string): this {
    Object.assign(this.schema, { type })
    return this
  }

  default(v: any): this {
    Object.assign(this.schema, { default: v })
    return this
  }

  oneOf(schemas: JsonSchema[]): this {
    Object.assign(this.schema, { oneOf: schemas })
    return this
  }

  allOf(schemas: JsonSchema[]): this {
    Object.assign(this.schema, { allOf: schemas })
    return this
  }

  instanceof(of: string): this {
    this.schema.instanceof = of
    return this
  }

  optional(optional = true): this {
    if (optional) {
      this.schema.optionalField = true
    } else {
      this.schema.optionalField = undefined
    }
    return this
  }

  /**
   * Produces a "clean schema object" without methods.
   * Same as if it would be JSON.stringified.
   */
  build(): SCHEMA_TYPE {
    return _sortObject(JSON.parse(JSON.stringify(this.schema)), JSON_SCHEMA_ORDER)
  }

  clone(): JsonSchemaAnyBuilder<T, SCHEMA_TYPE> {
    return new JsonSchemaAnyBuilder<T, SCHEMA_TYPE>(_deepCopy(this.schema))
  }
}

export class JsonSchemaNumberBuilder extends JsonSchemaAnyBuilder<number, JsonSchemaNumber> {
  constructor() {
    super({
      type: 'number',
    })
  }

  integer(): this {
    Object.assign(this.schema, { type: 'integer' })
    return this
  }

  multipleOf(multipleOf: number): this {
    Object.assign(this.schema, { multipleOf })
    return this
  }

  min(minimum: number): this {
    Object.assign(this.schema, { minimum })
    return this
  }

  exclusiveMin(exclusiveMinimum: number): this {
    Object.assign(this.schema, { exclusiveMinimum })
    return this
  }

  max(maximum: number): this {
    Object.assign(this.schema, { maximum })
    return this
  }

  exclusiveMax(exclusiveMaximum: number): this {
    Object.assign(this.schema, { exclusiveMaximum })
    return this
  }

  /**
   * Both ranges are inclusive.
   */
  range(minimum: number, maximum: number): this {
    Object.assign(this.schema, { minimum, maximum })
    return this
  }

  format(format: string): this {
    Object.assign(this.schema, { format })
    return this
  }

  int32 = (): this => this.format('int32')
  int64 = (): this => this.format('int64')
  float = (): this => this.format('float')
  double = (): this => this.format('double')
  unixTimestamp = (): this => this.format('unixTimestamp')
  unixTimestamp2000 = (): this => this.format('unixTimestamp2000')
  unixTimestampMillis = (): this => this.format('unixTimestampMillis')
  unixTimestampMillis2000 = (): this => this.format('unixTimestampMillis2000')
  utcOffset = (): this => this.format('utcOffset')
  utcOffsetHours = (): this => this.format('utcOffsetHours')
}

export class JsonSchemaStringBuilder extends JsonSchemaAnyBuilder<string, JsonSchemaString> {
  constructor() {
    super({
      type: 'string',
    })
  }

  pattern(pattern: string): this {
    Object.assign(this.schema, { pattern })
    return this
  }

  min(minLength: number): this {
    Object.assign(this.schema, { minLength })
    return this
  }

  max(maxLength: number): this {
    Object.assign(this.schema, { maxLength })
    return this
  }

  length(minLength: number, maxLength: number): this {
    Object.assign(this.schema, { minLength, maxLength })
    return this
  }

  format(format: string): this {
    Object.assign(this.schema, { format })
    return this
  }

  email = (): this => this.format('email')
  date = (): this => this.format('date')
  url = (): this => this.format('url')
  ipv4 = (): this => this.format('ipv4')
  ipv6 = (): this => this.format('ipv6')
  password = (): this => this.format('password')
  id = (): this => this.format('id')
  slug = (): this => this.format('slug')
  semVer = (): this => this.format('semVer')
  languageTag = (): this => this.format('languageTag')
  countryCode = (): this => this.format('countryCode')
  currency = (): this => this.format('currency')

  trim = (trim = true): this => this.transformModify('trim', trim)
  toLowerCase = (toLowerCase = true): this => this.transformModify('toLowerCase', toLowerCase)
  toUpperCase = (toUpperCase = true): this => this.transformModify('toUpperCase', toUpperCase)

  private transformModify(t: 'trim' | 'toLowerCase' | 'toUpperCase', add: boolean): this {
    if (add) {
      this.schema.transform = _uniq([...(this.schema.transform || []), t])
    } else {
      this.schema.transform = this.schema.transform?.filter(s => s !== t)
    }
    return this
  }

  // contentMediaType?: string
  // contentEncoding?: string
}

export class JsonSchemaObjectBuilder<T extends AnyObject> extends JsonSchemaAnyBuilder<
  T,
  JsonSchemaObject<T>
> {
  constructor() {
    super({
      type: 'object',
      properties: {} as T,
      required: [],
      additionalProperties: false,
    })
  }

  addProperties(props: { [k in keyof T]: JsonSchemaBuilder<T[k]> }): this {
    Object.entries(props).forEach(([k, builder]: [keyof T, JsonSchemaBuilder]) => {
      const schema = builder.build()
      if (!schema.optionalField) {
        this.schema.required.push(k)
      } else {
        schema.optionalField = undefined
      }
      this.schema.properties[k] = schema
    })

    this.required(this.schema.required) // ensure it's sorted and _uniq

    return this
  }

  /**
   * Ensures `required` is always sorted and _uniq
   */
  required(required: (keyof T)[]): this {
    Object.assign(this.schema, { required })
    this.schema.required = _uniq(required).sort()
    return this
  }

  addRequired(required: (keyof T)[]): this {
    return this.required([...this.schema.required, ...required])
  }

  minProps(minProperties: number): this {
    Object.assign(this.schema, { minProperties })
    return this
  }

  maxProps(maxProperties: number): this {
    Object.assign(this.schema, { maxProperties })
    return this
  }

  additionalProps(additionalProperties: boolean): this {
    Object.assign(this.schema, { additionalProperties })
    return this
  }

  baseDBEntity(): JsonSchemaObjectBuilder<T & BaseDBEntity> {
    Object.assign(this.schema.properties, {
      id: { type: 'string' },
      created: { type: 'number', format: 'unixTimestamp2000' },
      updated: { type: 'number', format: 'unixTimestamp2000' },
    })

    return this.addRequired(['id', 'created', 'updated']) as any
  }

  extend<T2 extends AnyObject>(s2: JsonSchemaObjectBuilder<T2>): JsonSchemaObjectBuilder<T & T2> {
    const builder = new JsonSchemaObjectBuilder<T & T2>()
    Object.assign(builder.schema, _deepCopy(this.schema))
    mergeJsonSchemaObjects(builder.schema, s2.schema)
    return builder
  }
}

export class JsonSchemaArrayBuilder<ITEM> extends JsonSchemaAnyBuilder<
  ITEM[],
  JsonSchemaArray<ITEM>
> {
  constructor(itemsSchema: JsonSchemaBuilder<ITEM>) {
    super({
      type: 'array',
      items: itemsSchema.build(),
    })
  }

  min(minItems: number): this {
    Object.assign(this.schema, { minItems })
    return this
  }

  max(maxItems: number): this {
    Object.assign(this.schema, { maxItems })
    return this
  }

  unique(uniqueItems: number): this {
    Object.assign(this.schema, { uniqueItems })
    return this
  }
}

export class JsonSchemaTupleBuilder<T extends any[]> extends JsonSchemaAnyBuilder<
  T,
  JsonSchemaTuple<T>
> {
  constructor(items: JsonSchemaBuilder[]) {
    super({
      type: 'array',
      items: items.map(b => b.build()),
      minItems: items.length,
      maxItems: items.length,
    })
  }
}
