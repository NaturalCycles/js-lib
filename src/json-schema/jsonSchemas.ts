import type { BaseDBEntity } from '../types.js'
import { jsonSchema } from './jsonSchemaBuilder.js'

export const baseDBEntityJsonSchema = jsonSchema.object<BaseDBEntity>({
  id: jsonSchema.string(),
  created: jsonSchema.unixTimestamp2000(),
  updated: jsonSchema.unixTimestamp2000(),
})
