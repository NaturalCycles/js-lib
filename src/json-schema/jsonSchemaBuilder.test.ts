import { AjvSchema } from '@naturalcycles/nodejs-lib'
import { expect, test } from 'vitest'
import type { BaseDBEntity } from '../index'
import { jsonSchema } from './jsonSchemaBuilder'
import { baseDBEntityJsonSchema } from './jsonSchemas'

interface Address {
  countryCode: string
  zip: string
  city: string
  address1: string
  address2?: string
  region?: string
  phone?: string
}

// interface AddressBM extends Address, BaseDBEntity {}
// interface AddressDBM extends Address, SavedDBEntity {}

const addressJsonSchema = jsonSchema.object<Address>({
  countryCode: jsonSchema.string().countryCode(),
  zip: jsonSchema.string().length(1, 40),
  city: jsonSchema.string(),
  address1: jsonSchema.string(),
  address2: jsonSchema.string().optional(),
  region: jsonSchema.string().optional(),
  phone: jsonSchema.string().optional(),
})

const addressBMJsonSchema = addressJsonSchema.extend(baseDBEntityJsonSchema)
const addressDBMJsonSchema = addressJsonSchema.extend(baseDBEntityJsonSchema)

// alternative

const addressBMJsonSchema2 = baseDBEntityJsonSchema.extend(addressJsonSchema)
const addressDBMJsonSchema2 = baseDBEntityJsonSchema.extend(addressJsonSchema)

// alternative 2
const addressBMJsonSchema3 = addressJsonSchema.extend(
  jsonSchema.object<BaseDBEntity>({
    id: jsonSchema.string(),
    created: jsonSchema.unixTimestamp2000(),
    updated: jsonSchema.unixTimestamp2000(),
  }),
)

test('simpleStringSchema', () => {
  const s = jsonSchema.string().build()

  expect(s).toMatchInlineSnapshot(`
    {
      "type": "string",
    }
  `)
})

test('addressSchema', () => {
  expect(addressJsonSchema.build()).toMatchSnapshot()
})

test('addressBMJsonSchema', () => {
  expect(addressBMJsonSchema.build()).toMatchSnapshot()
  expect(addressBMJsonSchema2.build()).toEqual(addressBMJsonSchema.build())
  expect(addressBMJsonSchema3.build()).toEqual(addressBMJsonSchema.build())
})

test('addressDBMJsonSchema', () => {
  expect(addressDBMJsonSchema.build()).toMatchSnapshot()
  expect(addressDBMJsonSchema2.build()).toEqual(addressDBMJsonSchema.build())
})

test('oneOf', () => {
  const s = jsonSchema.allOf([jsonSchema.string(), jsonSchema.string().countryCode()])
  expect(s.build()).toMatchInlineSnapshot(`
    {
      "allOf": [
        {
          "type": "string",
        },
        {
          "format": "countryCode",
          "type": "string",
        },
      ],
    }
  `)
})

test('order', () => {
  const s = addressDBMJsonSchema.$schemaDraft7().$id('AddressDBM').build()
  expect(Object.keys(s)).toMatchInlineSnapshot(`
    [
      "$schema",
      "$id",
      "type",
      "properties",
      "required",
      "additionalProperties",
    ]
  `)
})

test('buffer', () => {
  const s = jsonSchema.buffer()
  expect(s.build()).toMatchInlineSnapshot(`
    {
      "instanceof": "Buffer",
    }
  `)

  // const schema = AjvSchema.create(s) // this fails strangely!
  const schema = AjvSchema.create(s.build())
  schema.validate(Buffer.from('abc'))

  expect(schema.isValid('a b c' as any)).toBe(false)
})
