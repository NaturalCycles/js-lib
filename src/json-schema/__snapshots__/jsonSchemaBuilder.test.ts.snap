// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addressBMJsonSchema 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "address1": {
      "type": "string",
    },
    "address2": {
      "type": "string",
    },
    "city": {
      "type": "string",
    },
    "countryCode": {
      "format": "countryCode",
      "type": "string",
    },
    "created": {
      "format": "unixTimestamp2000",
      "type": "number",
    },
    "id": {
      "type": "string",
    },
    "phone": {
      "type": "string",
    },
    "region": {
      "type": "string",
    },
    "updated": {
      "format": "unixTimestamp2000",
      "type": "number",
    },
    "zip": {
      "maxLength": 40,
      "minLength": 1,
      "type": "string",
    },
  },
  "required": [
    "address1",
    "city",
    "countryCode",
    "zip",
  ],
  "type": "object",
}
`;

exports[`addressDBMJsonSchema 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "address1": {
      "type": "string",
    },
    "address2": {
      "type": "string",
    },
    "city": {
      "type": "string",
    },
    "countryCode": {
      "format": "countryCode",
      "type": "string",
    },
    "created": {
      "format": "unixTimestamp2000",
      "type": "number",
    },
    "id": {
      "type": "string",
    },
    "phone": {
      "type": "string",
    },
    "region": {
      "type": "string",
    },
    "updated": {
      "format": "unixTimestamp2000",
      "type": "number",
    },
    "zip": {
      "maxLength": 40,
      "minLength": 1,
      "type": "string",
    },
  },
  "required": [
    "address1",
    "city",
    "countryCode",
    "zip",
  ],
  "type": "object",
}
`;

exports[`addressSchema 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "address1": {
      "type": "string",
    },
    "address2": {
      "type": "string",
    },
    "city": {
      "type": "string",
    },
    "countryCode": {
      "format": "countryCode",
      "type": "string",
    },
    "phone": {
      "type": "string",
    },
    "region": {
      "type": "string",
    },
    "zip": {
      "maxLength": 40,
      "minLength": 1,
      "type": "string",
    },
  },
  "required": [
    "address1",
    "city",
    "countryCode",
    "zip",
  ],
  "type": "object",
}
`;
